src/Tools/SignRoslyn/RunSignUtil.cs

```diff
@@ -0,0 +1,360 @@
+                foreach (var part in package.GetParts())
+                {
+                    var relativeName = GetPartRelativeFileName(part);
+                    var vsixPart = vsixData.GetNestedBinaryPart(relativeName);
+                    if (!vsixPart.HasValue)
+                    {
+                        continue;
+                    }
+
+                    using (var stream = File.OpenRead(vsixPart.Value.BinaryName.FullPath))
+                    using (var partStream = part.GetStream(FileMode.Open, FileAccess.ReadWrite))
+                    {
+                        stream.CopyTo(partStream);
+                        partStream.SetLength(stream.Length);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the name of all VSIX which are nested inside this VSIX.
+        /// </summary>
+        private IEnumerable<string> GetNestedVsixRelativeNames(BinaryName vsixName)
+        {
+            return GetVsixPartRelativeNames(vsixName).Where(x => IsVsix(x));
+        }
+
+        private bool AreNestedVsixSigned(BinaryName vsixName, HashSet<string> signedSet)
+        {
+            foreach (var relativeName in GetNestedVsixRelativeNames(vsixName))
+            {
+                var name = Path.GetFileName(relativeName);
+                if (!signedSet.Contains(name))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Return all the assembly and VSIX contents nested in the VSIX
+        /// </summary>
+        private List<string> GetVsixPartRelativeNames(BinaryName vsixName)
+        {
+            var list = new List<string>();
+            using (var package = Package.Open(vsixName.FullPath, FileMode.Open, FileAccess.Read))
+            {
+                foreach (var part in package.GetParts())
+                {
+                    var name = GetPartRelativeFileName(part);
+                    list.Add(name);
+                }
+            }
+
+            return list;
+        }
+
+        private Dictionary<BinaryName, VsixData> VerifyBeforeSign()
+        {
+            var allGood = true;
+            var map = VerifyBinariesBeforeSign(ref allGood);
+            var vsixDataMap = VerifyVsixContentsBeforeSign(map, ref allGood);
+
+            if (!allGood)
+            {
+                throw new Exception("Errors validating the state before signing");
+            }
+
+            return vsixDataMap;
+        }
+
+        /// <summary>
+        /// Validate all of the binaries which are specified to be signed exist on disk.  Compute their
+        /// checksums at this time so we can use it for VSIX content validation.
+        /// </summary>
+        private Dictionary<string, BinaryName> VerifyBinariesBeforeSign(ref bool allGood)
+        {
+            var checksumToNameMap = new Dictionary<string, BinaryName>(StringComparer.Ordinal);
```

Comment:
In this tool the act of signing dominates performance.  On a good run it takes 5 minutes to do a sign operation, on a bad run the network issues can stretch it to nearly an hour.  Hence parallelization of this complicates the program for no real performance win.